File: Leetcode/Python Solutions/Strings/LongestCommonSubsequence.py
"""
LeetCode Problem: 1143. Longest Common Subsequence
Link: https://leetcode.com/problems/longest-common-subsequence/
Written by: Mostofa Adib Shakib
Language: Python

"""


"""
Recursive solution - Inefficient
Time Complexity: O(2^n)
Space Complexity: O(1)
"""

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if not text1 or not text2:
            return 0

        if text1[-1] == text2[-1]:
            text1 = text1[:-1]
            text2 = text2[:-1]
            return 1 + self.longestCommonSubsequence(text1, text2)
        else:
            return max(self.longestCommonSubsequence(text1, text2[:-1]), self.longestCommonSubsequence(text1[:-1], text2))


"""
Dynamic Programming
Time Complexity: O(n^2)
Space Complexity: O(n^2)
"""

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if not text1 or not text2: return 0
        
        n = len(text1)
        m = len(text2)

        dp= [[0 for i in range(m+1)] for j in range(n+1)]

        for i in range(0, n):
            for j in range(0, m):
                if text1[i] == text2[j]:
                    dp[i+1][j+1] = 1 + dp[i][j]
                else:
                    dp[i+1][j+1] = max(dp[i][j+1], dp[i+1][j])
        
        return dp[-1][-1]
    

File: Leetcode/Java Solutions/longestcommonsubsequence.java
/*
LeetCode Problem: 1143. Longest Common Subsequence
Link: https://leetcode.com/problems/longest-common-subsequence/
Written by: Mostofa Adib Shakib
Language: Java
Time Complexity: O(n*m)
Space Complexity: O(n*m)
*/

class Solution {
    public int[][] dp;
    
    public int longestCommonSubsequence(String text1, String text2) {
        int text1_length = text1.length();    // column
        int text2_length = text2.length();    // row
        
        dp = new int[text2_length + 1][text1_length + 1];
        
        for (int i = 0; i < text2_length+1; i++) {
            for (int j = 0; j < text1_length+1; j++) {
                if (i == 0 || j == 0){
                    dp[i][j] = 0;
                }
                else if ( text2.charAt(i-1) == text1.charAt(j-1)  ){
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else{
                    dp[i][j] = Math.max(dp[i][j-1], dp[i-1][j]) ;
                }
            }
        }
        return dp[text2_length][text1_length];
    }
}